spark/sql-13/src/main/scala/org/elasticsearch/spark/sql/DataFrameValueWriter.scala

```diff
@@ -64,11 +60,25 @@ class DataFrameValueWriter(writeUnknownTypes: Boolean = false) extends Filtering
     writeNullValues = settings.getDataFrameWriteNullValues
   }
 
-  override def write(value: (Row, StructType), generator: Generator): Result = {
-    val row = value._1
-    val schema = value._2
+  override def write(value: Any, generator: Generator): Result = {
+    value match {
+      case Tuple2(row, schema: StructType) =>
+        writeStruct(schema, row, generator)
+      case seq: Seq[Row] =>
+        writeArray(seq, generator)
+    }
+  }
 
-    writeStruct(schema, row, generator)
+  private[spark] def writeArray(value: Seq[Row], generator: Generator): Result = {
+    if (value.nonEmpty) {
+      val schema = value.head.schema
+      val result = write(DataTypes.createArrayType(schema), value, generator)
+      if (!result.isSuccesful) {
+        return handleUnknown(value, generator)
+      }
+      Result.SUCCESFUL()
+    }
+    Result.FAILED()
```

Comment:
Is an empty array a failure scenario here for sure?

Reply:
I can't remember for sure at this point, but I think that my thinking was that in this case we had been asked to write something, but hadn't actually written an array, so that's failure. Maybe that doesn't make sense though. I'll take a look at what's done with that result. Also I can't remember why I'm not just writing an empty array here.

Reply:
It actually turns out that we ignore this result. See https://github.com/elastic/elasticsearch-hadoop/blob/master/mr/src/main/java/org/elasticsearch/hadoop/serialization/bulk/AbstractBulkFactory.java#L153. We probably ought to be throwing an exception there. But an empty array actually is a failure (I think). I'm about to write more in the comment below.